ratelimits/limiter.go

```diff
@@ -277,31 +277,34 @@ func (l *Limiter) BatchSpend(ctx context.Context, txns []Transaction) (*Decision
 	batchDecision := allowedDecision
 	newBuckets := make(map[string]time.Time)
 	incrBuckets := make(map[string]increment)
+	staleBuckets := make(map[string]time.Time)
 	txnOutcomes := make(map[Transaction]string)
 
 	for _, txn := range batch {
-		tat, bucketExists := tats[txn.bucketKey]
+		storedTAT, bucketExists := tats[txn.bucketKey]
+		effectiveTAT := storedTAT
 		if !bucketExists {
 			// First request from this client.
-			tat = l.clk.Now()
+			effectiveTAT = l.clk.Now()
 		}
 
-		d := maybeSpend(l.clk, txn, tat)
+		d := maybeSpend(l.clk, txn, effectiveTAT)
```

Comment:
You're absolutely right. I think this can just be simplified to:

```go
	for _, txn := range batch {
		storedTAT, bucketExists := tats[txn.bucketKey]
		d := maybeSpend(l.clk, txn, storedTAT)

                ...

		if d.allowed && (storedTAT != d.newTAT) && txn.spend {
			if !bucketExists {
				newBuckets[txn.bucketKey] = d.newTAT
			} else if storedTAT.After(l.clk.Now()) {
				incrBuckets[txn.bucketKey] = increment{
					cost: time.Duration(txn.cost * txn.limit.emissionInterval),
					ttl:  time.Duration(txn.limit.burstOffset),
				}
			} else {
				staleBuckets[txn.bucketKey] = d.newTAT
			}
		}
```